Paintroid/src/main/java/org/catrobat/paintroid/iotasks/SaveImage.kt

```diff
@@ -0,0 +1,137 @@
+/*
+ * Paintroid: An image manipulation application for Android.
+ * Copyright (C) 2010-2015 The Catrobat Team
+ * (<http://developer.catrobat.org/credits>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.catrobat.paintroid.iotasks
+
+import android.content.ContentResolver
+import android.net.Uri
+import android.util.Log
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import org.catrobat.paintroid.FileIO
+import org.catrobat.paintroid.common.Constants
+import org.catrobat.paintroid.tools.Workspace
+import java.io.IOException
+import java.lang.ref.WeakReference
+
+class SaveImage(
+    activity: SaveImageCallback,
+    private val requestCode: Int,
+    private val workspace: Workspace,
+    private var uri: Uri?,
+    private val saveAsCopy: Boolean,
+    private val scopeIO: CoroutineScope
+) {
+    private val callbackRef: WeakReference<SaveImageCallback> = WeakReference(activity)
+
+    companion object {
+        private val TAG = SaveImage::class.java.simpleName
+    }
+
+    fun execute() {
```

Comment:
Please refactor.

Reply:
Could please take a quick look at it again? I've tried to make new functions and reduced the nested lines :)